File: apps/web/src/lib/pair-device.ts
/**
 * Pair Device — wallet-to-wallet transfer of session DEKs over an
 * untrusted cloud mailbox.
 *
 * The flow is asymmetric. There is a **receiver** (a fresh browser that
 * needs to inherit the user's existing chat sessions) and a **sender**
 * (a browser already holding session DEKs in its
 * {@link SessionVault}).
 *
 * 1. The receiver calls {@link createReceiverHandshake} to generate an
 *    ephemeral X25519 keypair, a high-entropy mailbox id, and a small
 *    JSON descriptor that the UI encodes into a QR code.
 * 2. The sender calls {@link sendHandshake}, passing the receiver's
 *    QR-decoded descriptor, the user's vault, and a sealed-envelope
 *    signing callback (Turnkey in production, local key in tests). It
 *    builds an envelope addressed to the receiver's ephemeral pubkey
 *    and POSTs it to `/api/devices/handshake`.
 * 3. The receiver calls {@link awaitHandshake}, which polls the cloud
 *    until the envelope arrives, opens it with the ephemeral X25519
 *    secret, verifies the sender's signature against the **expected
 *    sender DID** (the receiver UI must show this DID and let the user
 *    confirm before importing — that's the only step where the cloud
 *    cannot impersonate), and imports each DEK into the receiver's
 *    vault.
 *
 * The cloud sees only opaque ciphertext + an unguessable id. See
 * `crates/thumper-cloud/src/routes/handshake.rs` for the server side
 * and `crates/said-envelope/src/lib.rs` for the wire format.
 */

import { ed25519, x25519 } from "@noble/curves/ed25519";

import {
  RecipientKind,
  didKeyFromVerifying,
  open as openEnvelope,
  seal as sealEnvelope,
  type SealOptions,
} from "./envelope";
import {
  SessionVault,
  VaultRecipientKind,
  type VaultRecipientKindByte,
} from "./session-vault";

const THUMPER_API_BASE =
  (typeof process !== "undefined" && process.env?.NEXT_PUBLIC_THUMPER_API_URL) ||
  "http://localhost:3000";

const HANDSHAKE_AD = new TextEncoder().encode("ghola/pair-device-v1");

// Server-side validate_id requires ≥16 decoded bytes; we go a little
// over to leave room for entropy growth.
const HANDSHAKE_ID_BYTES = 24;

const POLL_INTERVAL_MS = 1000;
const POLL_TIMEOUT_MS = 110_000; // a bit under the server-side TTL

// ── Wire payload (envelope plaintext) ───────────────────────────────────

interface DekTransferPayload {
  /** Schema version. Bumps when the inner format changes. */
  v: 1;
  sessions: Array<{
    sessionId: string;
    /** Base64-encoded 32-byte DEK. */
    dekB64: string;
    recipientKind: VaultRecipientKindByte;
    /** Unix-ms when the original DEK was generated, for audit only. */
    createdAt: number;
  }>;
}

// ── Receiver: descriptor + await ────────────────────────────────────────

export interface HandshakeDescriptor {
  /** Schema version. */
  v: 1;
  /** The mailbox id the sender will POST to. */
  id: string;
  /** Receiver's ephemeral X25519 pubkey, base64url-no-pad. */
  ephemPubB64: string;
  /** The DID of the **expected sender** (the user's own DID). UIs MUST
   *  show this to the receiving user before any import — it's the only
   *  trust anchor the receiver has against a malicious cloud
   *  substituting a different envelope. */
  expectedSenderDid: string;
}

export interface ReceiverHandshake {
  /** Encoded descriptor — embed in a QR code or copy/paste fallback. */
  descriptor: HandshakeDescriptor;
  /** Receiver's ephemeral X25519 secret. Held in memory only; if the
   *  user navigates away before {@link awaitHandshake} resolves, the
   *  envelope cannot be recovered. */
  ephemSecret: Uint8Array;
}

/**
 * Generate a fresh receiver handshake. `userDid` is the
 * **expected sender DID** — the user's wallet DID on the existing
 * device — which the receiver pins so the cloud can't substitute a
 * different sender.
 */
export function createReceiverHandshake(userDid: string): ReceiverHandshake {
  const idBytes = new Uint8Array(HANDSHAKE_ID_BYTES);
  crypto.getRandomValues(idBytes);
  const id = b64urlEncode(idBytes);

  const ephemSecret = x25519.utils.randomPrivateKey();
  const ephemPub = x25519.getPublicKey(ephemSecret);

  return {
    descriptor: {
      v: 1,
      id,
      ephemPubB64: b64urlEncode(ephemPub),
      expectedSenderDid: userDid,
    },
    ephemSecret,
  };
}

export interface AwaitHandshakeOptions {
  receiver: ReceiverHandshake;
  vault: SessionVault;
  /** Signal to abort polling early (e.g. user closed the modal). */
  signal?: AbortSignal;
}

export interface AwaitHandshakeResult {
  /** How many session DEKs were imported into the vault. */
  imported: number;
  /** The DID of the sender (always equals
   *  `receiver.descriptor.expectedSenderDid` if verification passed). */
  senderDid: string;
}

/**
 * Poll the mailbox until the sender's envelope arrives. Verifies the
 * signature, decrypts, and imports every DEK into the vault.
 *
 * Throws if the envelope's `senderDid` does not match the receiver's
 * pinned `expectedSenderDid` — i.e. the cloud handed us an envelope
 * from someone other than the user's own wallet.
 */
export async function awaitHandshake(
  opts: AwaitHandshakeOptions,
): Promise<AwaitHandshakeResult> {
  const { receiver, vault, signal } = opts;
  const id = receiver.descriptor.id;
  const deadline = Date.now() + POLL_TIMEOUT_MS;

  while (Date.now() < deadline) {
    if (signal?.aborted) throw new Error("handshake aborted");
    const wire = await pollOnce(id, signal);
    if (wire) {
      return importEnvelope(wire, receiver, vault);
    }
    await sleep(POLL_INTERVAL_MS, signal);
  }
  throw new Error("handshake timed out");
}

async function importEnvelope(
  wire: Uint8Array,
  receiver: ReceiverHandshake,
  vault: SessionVault,
): Promise<AwaitHandshakeResult> {
  const opened = await openEnvelope(wire, receiver.ephemSecret);

  if (opened.senderDid !== receiver.descriptor.expectedSenderDid) {
    throw new Error(
      `unexpected sender DID: got ${opened.senderDid}, expected ${receiver.descriptor.expectedSenderDid}`,
    );
  }

  const json = new TextDecoder().decode(opened.plaintext);
  const payload = JSON.parse(json) as DekTransferPayload;
  if (payload.v !== 1 || !Array.isArray(payload.sessions)) {
    throw new Error("invalid Pair-Device payload format");
  }

  for (const s of payload.sessions) {
    const dek = b64urlDecode(s.dekB64);
    if (dek.length !== 32) {
      throw new Error(`invalid DEK length for session ${s.sessionId}`);
    }
    await vault.importSessionDek(s.sessionId, dek, s.recipientKind);
  }

  return { imported: payload.sessions.length, senderDid: opened.senderDid };
}

async function pollOnce(
  id: string,
  signal?: AbortSignal,
): Promise<Uint8Array | null> {
  const url = `${THUMPER_API_BASE}/api/devices/handshake/${encodeURIComponent(id)}`;
  const res = await fetch(url, { method: "GET", signal });
  if (res.status === 404) return null;
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new Error(`handshake poll failed (${res.status}): ${body}`);
  }
  const data = (await res.json()) as { envelope_b64: string };
  return b64urlDecode(data.envelope_b64);
}

// ── Sender ─────────────────────────────────────────────────────────────

export interface SendHandshakeOptions {
  /** Descriptor scanned from the receiver's QR code. */
  descriptor: HandshakeDescriptor;
  /** Sender's unlocked vault. The DEKs returned by
   *  {@link SessionVault.listSessions} are wrapped onto the wire. */
  vault: SessionVault;
  /** Sender's `did:key:z…`. MUST equal `descriptor.expectedSenderDid`
   *  on the receiving side, otherwise the receiver rejects the
   *  envelope. */
  senderDid: string;
  /** Sealed-envelope signing callback. Production passes Turnkey;
   *  tests pass a local Ed25519 signer
   *  (see {@link envelope.localEd25519Signer}). */
  signBody: SealOptions["signBody"];
  /** Optional whitelist; if omitted, every session in the vault is
   *  transferred. */
  sessionIds?: string[];
}

/**
 * Build a sealed envelope from the sender's session DEKs and POST it to
 * the cloud mailbox. Returns the count transferred.
 */
export async function sendHandshake(opts: SendHandshakeOptions): Promise<number> {
  const { descriptor, vault, senderDid, signBody, sessionIds } = opts;

  if (descriptor.v !== 1) {
    throw new Error(`unsupported handshake descriptor version: ${descriptor.v}`);
  }

  const ephemPub = b64urlDecode(descriptor.ephemPubB64);
  if (ephemPub.length !== 32) throw new Error("descriptor ephemPub is wrong length");

  const allSessions = await vault.listSessions();
  const wanted = sessionIds
    ? allSessions.filter((s) => sessionIds.includes(s.sessionId))
    : allSessions;

  const payload: DekTransferPayload = {
    v: 1,
    sessions: [],
  };
  for (const s of wanted) {
    const dek = await vault.getSessionDek(s.sessionId);
    if (!dek) continue; // raced with deletion — drop it
    payload.sessions.push({
      sessionId: s.sessionId,
      dekB64: b64urlEncode(dek),
      recipientKind: s.recipientKind,
      createdAt: s.createdAt,
    });
  }

  const plaintext = new TextEncoder().encode(JSON.stringify(payload));

  // recipientId is the receiver's ephem pub serialized as
  // `ephem-x25519:<base64>`. It's used as part of the HKDF info string,
  // not as the X25519 ECDH input — that's the raw pub bytes. Keeping
  // a stable string lets the receiver bind the DEK to the descriptor it
  // generated.
  const recipientId = `ephem-x25519:${descriptor.ephemPubB64}`;

  const wire = await sealEnvelope({
    senderDid,
    recipientId,
    recipientX25519: ephemPub,
    kind: RecipientKind.PeerDid,
    associatedData: HANDSHAKE_AD,
    plaintext,
    signBody,
  });

  const res = await fetch(`${THUMPER_API_BASE}/api/devices/handshake`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      id: descriptor.id,
      envelope_b64: b64urlEncode(wire),
    }),
  });
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new Error(`handshake POST failed (${res.status}): ${body}`);
  }

  return payload.sessions.length;
}

// ── helpers ────────────────────────────────────────────────────────────

function b64urlEncode(bytes: Uint8Array): string {
  let bin = "";
  for (let i = 0; i < bytes.length; i++) bin += String.fromCharCode(bytes[i]);
  // btoa produces standard base64; convert to URL-safe no-pad.
  return btoa(bin).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function b64urlDecode(s: string): Uint8Array {
  const padded = s.replace(/-/g, "+").replace(/_/g, "/");
  const pad = padded.length % 4 === 0 ? padded : padded + "=".repeat(4 - (padded.length % 4));
  const bin = atob(pad);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    if (signal?.aborted) return reject(new Error("aborted"));
    const t = setTimeout(() => resolve(), ms);
    signal?.addEventListener(
      "abort",
      () => {
        clearTimeout(t);
        reject(new Error("aborted"));
      },
      { once: true },
    );
  });
}

// `_unused` so eslint doesn't flag the imports purely for type usage.
type _unused = typeof ed25519 | VaultRecipientKindByte;
void VaultRecipientKind; // referenced for re-export discoverability

File: android/app/src/main/java/xyz/ghola/app/crypto/PairDevice.kt
package xyz.ghola.app.crypto

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.security.SecureRandom
import java.util.concurrent.TimeUnit

/**
 * Pair Device — wallet-to-wallet transfer of session DEKs over an
 * untrusted cloud mailbox. Mirror of `apps/web/src/lib/pair-device.ts`.
 *
 * Same-platform only (Android↔Android in v0.3) — see plan, Out of scope §1.
 * Web↔Android pairing requires aligning sender-DID conventions; tracked
 * for a follow-up.
 */
object PairDevice {

    private const val HANDSHAKE_ID_BYTES = 24
    private const val POLL_INTERVAL_MS = 1_000L
    private const val POLL_TIMEOUT_MS = 110_000L
    private val HANDSHAKE_AD = "ghola/pair-device-v1".toByteArray(Charsets.UTF_8)
    private val JSON_MEDIA = "application/json; charset=utf-8".toMediaType()
    private val rng = SecureRandom()

    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .writeTimeout(15, TimeUnit.SECONDS)
        .build()

    /** What the receiver renders into a QR code for the sender to scan. */
    data class HandshakeDescriptor(
        val v: Int,
        val id: String,
        val ephemPubB64: String,
        val expectedSenderDid: String,
    ) {
        fun toJson(): String = JSONObject().apply {
            put("v", v)
            put("id", id)
            put("ephemPubB64", ephemPubB64)
            put("expectedSenderDid", expectedSenderDid)
        }.toString()

        companion object {
            fun fromJson(s: String): HandshakeDescriptor {
                val o = JSONObject(s)
                return HandshakeDescriptor(
                    v = o.getInt("v"),
                    id = o.getString("id"),
                    ephemPubB64 = o.getString("ephemPubB64"),
                    expectedSenderDid = o.getString("expectedSenderDid"),
                )
            }
        }
    }

    /** Held by the receiving device until the sender's envelope arrives. */
    data class ReceiverHandshake(
        val descriptor: HandshakeDescriptor,
        /** 32-byte ephemeral X25519 secret. In-memory only. */
        val ephemSecret: ByteArray,
    ) {
        fun zeroize() { for (i in ephemSecret.indices) ephemSecret[i] = 0 }
    }

    data class AwaitResult(val imported: Int, val senderDid: String)

    // ── Receiver ────────────────────────────────────────────────────────

    /**
     * Generate a fresh receiver handshake. `expectedSenderDid` MUST be
     * the user's own wallet DID — it's the only trust anchor we have
     * against a malicious cloud substituting an envelope from someone
     * else.
     */
    fun createReceiverHandshake(expectedSenderDid: String): ReceiverHandshake {
        val idBytes = ByteArray(HANDSHAKE_ID_BYTES).also { rng.nextBytes(it) }
        val id = b64urlEncode(idBytes)

        val ephemPriv = X25519PrivateKeyParameters(rng)
        val ephemPub = ephemPriv.generatePublicKey().encoded

        return ReceiverHandshake(
            descriptor = HandshakeDescriptor(
                v = 1,
                id = id,
                ephemPubB64 = b64urlEncode(ephemPub),
                expectedSenderDid = expectedSenderDid,
            ),
            ephemSecret = ephemPriv.encoded,
        )
    }

    /**
     * Poll the cloud mailbox until the sender's envelope arrives, then
     * verify, decrypt, and import each DEK into [vault]. Throws if the
     * sender DID doesn't match the receiver's pinned expected DID — a
     * malicious cloud cannot bypass that.
     *
     * Blocks the calling thread; run from a coroutine scope on
     * Dispatchers.IO.
     */
    @Throws(IOException::class)
    fun awaitHandshake(
        baseUrl: String,
        receiver: ReceiverHandshake,
        vault: VaultStore,
    ): AwaitResult {
        val deadline = System.currentTimeMillis() + POLL_TIMEOUT_MS
        while (System.currentTimeMillis() < deadline) {
            val wire = pollOnce(baseUrl, receiver.descriptor.id)
            if (wire != null) {
                return importEnvelope(wire, receiver, vault)
            }
            Thread.sleep(POLL_INTERVAL_MS)
        }
        throw IOException("pair-device handshake timed out after ${POLL_TIMEOUT_MS / 1000}s")
    }

    private fun importEnvelope(
        wire: ByteArray,
        receiver: ReceiverHandshake,
        vault: VaultStore,
    ): AwaitResult {
        val opened = Envelope.open(wire, receiver.ephemSecret)
        if (opened.senderDid != receiver.descriptor.expectedSenderDid) {
            throw IOException(
                "unexpected sender DID: got ${opened.senderDid}, " +
                    "expected ${receiver.descriptor.expectedSenderDid}"
            )
        }
        val payloadStr = String(opened.plaintext, Charsets.UTF_8)
        val payload = JSONObject(payloadStr)
        if (payload.optInt("v", 0) != 1) {
            throw IOException("invalid Pair-Device payload version")
        }
        val sessions = payload.optJSONArray("sessions") ?: JSONArray()
        var imported = 0
        for (i in 0 until sessions.length()) {
            val s = sessions.optJSONObject(i) ?: continue
            val sessionId = s.optString("sessionId").takeIf { it.isNotEmpty() } ?: continue
            val dekB64 = s.optString("dekB64").takeIf { it.isNotEmpty() } ?: continue
            val dek = b64urlDecode(dekB64)
            if (dek.size != 32) throw IOException("invalid DEK length for $sessionId")
            val kindByte = (s.optInt("recipientKind", 0) and 0xFF).toByte()
            val kind = try {
                Envelope.RecipientKind.fromByte(kindByte)
            } catch (_: Exception) {
                Envelope.RecipientKind.SelfRecipient
            }
            vault.importSessionDek(sessionId, dek, kind)
            imported++
        }
        return AwaitResult(imported = imported, senderDid = opened.senderDid)
    }

    private fun pollOnce(baseUrl: String, id: String): ByteArray? {
        val req = Request.Builder()
            .url("$baseUrl/api/devices/handshake/$id")
            .get()
            .build()
        return httpClient.newCall(req).execute().use { resp ->
            when (resp.code) {
                404 -> null
                in 200..299 -> {
                    val body = resp.body?.string() ?: return@use null
                    val obj = JSONObject(body)
                    b64urlDecode(obj.getString("envelope_b64"))
                }
                else -> {
                    val msg = resp.body?.string() ?: ""
                    throw IOException("handshake poll failed (${resp.code}): $msg")
                }
            }
        }
    }

    // ── Sender ──────────────────────────────────────────────────────────

    /**
     * Build a sealed handshake envelope from the sender's session DEKs
     * and POST it to the cloud mailbox.
     *
     * The body is signed with the sender's wallet via [walletSigner] —
     * one MWA popup is acceptable in this rare flow because the receiver
     * pins `expectedSenderDid` to the wallet DID and accepts only that
     * identity. (Chat-message envelopes use the cached chat-sign seed
     * instead; that asymmetry is the source of the "no cross-platform
     * v0.3" constraint, see the approved plan.)
     */
    @Throws(IOException::class)
    fun sendHandshake(
        baseUrl: String,
        descriptor: HandshakeDescriptor,
        vault: VaultStore,
        senderWalletDid: String,
        walletSigner: Envelope.Ed25519BodySigner,
        sessionIds: List<String>? = null,
    ): Int {
        require(descriptor.v == 1) { "unsupported handshake descriptor version: ${descriptor.v}" }
        val ephemPub = b64urlDecode(descriptor.ephemPubB64)
        require(ephemPub.size == 32) { "descriptor ephemPub must be 32 bytes" }

        val all = vault.listSessions()
        val wanted = if (sessionIds == null) all
        else all.filter { it.sessionId in sessionIds }

        val sessionsJson = JSONArray()
        for (s in wanted) {
            val dek = try {
                vault.getOrCreateSessionDek(s.sessionId, s.recipientKind)
            } catch (_: Exception) {
                continue // raced with deletion or vault locked — skip
            }
            sessionsJson.put(JSONObject().apply {
                put("sessionId", s.sessionId)
                put("dekB64", b64urlEncode(dek))
                put("recipientKind", s.recipientKind.byte.toInt() and 0xFF)
                put("createdAt", System.currentTimeMillis())
            })
        }
        val payload = JSONObject().apply {
            put("v", 1)
            put("sessions", sessionsJson)
        }
        val plaintext = payload.toString().toByteArray(Charsets.UTF_8)

        // recipientId is "ephem-x25519:<base64>" — used only as the HKDF
        // info string at DEK derivation, NOT as the X25519 ECDH input
        // (that's the raw pub bytes). Same convention as web.
        val recipientId = "ephem-x25519:${descriptor.ephemPubB64}"

        val wire = Envelope.seal(
            Envelope.SealParams(
                senderDid = senderWalletDid,
                kind = Envelope.RecipientKind.PeerDid,
                recipientId = recipientId,
                recipientX25519 = ephemPub,
                associatedData = HANDSHAKE_AD,
                plaintext = plaintext,
                signBody = walletSigner,
            ),
        )

        val body = JSONObject().apply {
            put("id", descriptor.id)
            put("envelope_b64", b64urlEncode(wire))
        }.toString().toRequestBody(JSON_MEDIA)
        val req = Request.Builder()
            .url("$baseUrl/api/devices/handshake")
            .post(body)
            .build()
        httpClient.newCall(req).execute().use { resp ->
            if (!resp.isSuccessful) {
                val msg = resp.body?.string() ?: ""
                throw IOException("handshake POST failed (${resp.code}): $msg")
            }
        }
        return sessionsJson.length()
    }

    // ── b64url helpers (no-pad, matches the web client) ─────────────────

    private val b64UrlEncoder = java.util.Base64.getUrlEncoder().withoutPadding()
    private val b64UrlDecoder = java.util.Base64.getUrlDecoder()

    private fun b64urlEncode(bytes: ByteArray): String = b64UrlEncoder.encodeToString(bytes)
    private fun b64urlDecode(s: String): ByteArray = b64UrlDecoder.decode(s)
}
